### PR TITLE
fix:  change email update flow to return both ? messages and # messages

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

We amend the error messages to return both ? messages and # messages when using the PKCE flow for backward compatibility with the client libraries. The client libraries will be able to fetch error message from error fragments with this change.

### CODE REVIEW

File: internal/api/verify.go [hunk 1]
```diff
@@ -364,13 +373,25 @@ func (a *API) prepErrorRedirectURL(err *HTTPError, w http.ResponseWriter, r *htt
 	}
 	q.Set("error_code", strconv.Itoa(err.Code))
 	q.Set("error_description", err.Message)
-	return rurl + "#" + q.Encode()
+	if flowType == models.PKCEFlow {
```
[the remark this answers is not in the document]
Comment (J0) [reply internal/api/verify.go:376]: Yup, this is intended as we don't wish to do so for the other error redirect URLs (OAuth). Just the email change redirects. 

The hash and query param is appended here: https://github.com/supabase/gotrue/pull/1129/files#diff-e47b259bf17912f7ca1f5163ccbfb8a8c23da6c399412ce5aef78b8b684df5e5R166

[the remark this answers is not in the document]
Comment (silentworks) [reply internal/api/verify.go:376]: These are the lines I'm having contention with as its stating either `?` or `#` and not both. https://github.com/supabase/gotrue/pull/1129/files#diff-e47b259bf17912f7ca1f5163ccbfb8a8c23da6c399412ce5aef78b8b684df5e5R377-R379

[the remark this answers is not in the document]
Comment (J0) [reply internal/api/verify.go:376]: Ah okay, I see where the confusion might be coming from. We currently do [append both parameters if there's a PKCE flow](https://github.com/supabase/gotrue/pull/1129/files#diff-e47b259bf17912f7ca1f5163ccbfb8a8c23da6c399412ce5aef78b8b684df5e5R128)

However, this is currently done  [in `verifyGet`  and treated as a special case](https://github.com/supabase/gotrue/pull/1129/files#diff-e47b259bf17912f7ca1f5163ccbfb8a8c23da6c399412ce5aef78b8b684df5e5R128) which might be confusing. Let me refactor to make it clearer

File: internal/api/verify.go [hunk 2]
```diff
@@ -364,13 +365,29 @@ func (a *API) prepErrorRedirectURL(err *HTTPError, w http.ResponseWriter, r *htt
...
+func (a *API) prepRedirectURL(message string, rurl string, flowType models.FlowType) string {
 	q := url.Values{}
 	q.Set("message", message)
-	return rurl + "#" + q.Encode()
+	var fragmentDelimiter string
+
+	if flowType == models.PKCEFlow {
+		fragmentDelimiter = "?"
+		// For backward compatibility with client library, we append implict error redirect as well
+		return a.prepRedirectURL(message, rurl+fragmentDelimiter+q.Encode(), models.ImplicitFlow)
```
Comment (hf) [inline internal/api/verify.go:386]: These lines:

```
		return a.prepRedirectURL(message, rurl+fragmentDelimiter+q.Encode(), models.ImplicitFlow)
```

Seem to repeat a bit? Maybe extract them into a function we can test?

**Review by hf — APPROVED** [review]

**Review by kangmingtay — COMMENTED** [review]
just to clarify, are we planning to capture the error messages in the query params for all redirected URLs or just for the PKCE flows?

iirc, we wanted to do it for all, but we're also keep the errors in the fragment for backward compatibility sake

**Comment by J0** [comment]
Hm my understanding was that we are doing it only for the PKCE flows.  Which other redirect urls did you have in mind? We can add it in separate PR if all are agreeable

**Comment by kangmingtay** [comment]
yeah sure, it can be a separate PR, it would be for any endpoint that does a http redirect with the error message appended to the url as a fragment 

i think we should standardise all of these endpoints to append the error message as a query param instead

**Comment by kangmingtay** [comment]
btw can you also test what happens if the `redirect_to` url contains a query param? will the error messages added remove those query params or will they just be appended? also what happens if the `redirect_to` url contains query params with the same keywords as those we use in gotrue when an error is returned?

**Comment by J0** [comment]
Yes, just realized that we need to factor that but also wondering what to do with error messages

**Comment by J0** [comment]
Looks like the URL package strips the trailing / which may affect the URL redirect allow list, looking for a workaround.

File: internal/api/verify.go [hunk 3]
```diff
@@ -392,13 +404,25 @@ func (a *API) prepErrorRedirectURL(err *HTTPError, w http.ResponseWriter, r *htt
...
+	u, perr := url.Parse(rurl)
+	if perr != nil {
+		return "", perr
+	}
+	q := u.Query()
+	if flowType == models.PKCEFlow {
+		q.Set("message", message)
+	}
+	u.RawQuery = q.Encode()
+	u.Fragment = fmt.Sprintf("message=%s", url.QueryEscape(message))
```
Comment (J0) [inline internal/api/verify.go:424]: We don't  use `q.Encode()` as there might be existing query params in the url which we don't want to pass into the fragment

File: internal/api/verify.go [hunk 4]
```diff
@@ -381,7 +387,13 @@ func (a *API) smsVerify(r *http.Request, ctx context.Context, conn *storage.Conn
 	return user, nil
 }
 
-func (a *API) prepErrorRedirectURL(err *HTTPError, w http.ResponseWriter, r *http.Request, rurl string) string {
+func (a *API) prepErrorRedirectURL(err *HTTPError, w http.ResponseWriter, r *http.Request, rurl string, flowType models.FlowType) (string, error) {
+	u, perr := url.Parse(rurl)
+	if perr != nil {
+		return "", err
+	}
+	// Don't preserve query params for error
```
Comment (J0) [inline internal/api/verify.go:395]: Query params should be preserved in case of oauth error

**Pull request merged**